Add new cargo.md tutorial and standalone test projects

sffc opened the pull request:
_I don't expect reviews on PRs I send this week until January_

Fixes #2853
Fixes #2752

I'm not sure how to best test these examples to make sure they don't go stale. Putting them on CI is expensive because everything needs to be downloaded fresh, and they could be flaky since the deps are pulled from crates.io.

Manishearth commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,114 @@
+# Setting up your Cargo.toml file for ICU4X
```
thought: increasingly of the opinion that we should be making an MdBook that contains an organized set of tutorials as well as deeper docs on e.g. stability policies

sffc replied:
Agreed; the docs folder gives us a lot of material we can use to start it

Manishearth commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,114 @@
+
+## Cargo.toml with Baked Provider
+
+If you wish to bake in data auto-generated in your build.rs, you need to add additional dependencies as follows:
```
issue: I'm a bit wary of recommending people call datagen (which is super heavy) in a build script without adding all the caveats that it is super heavy and linking to a dedicated tutorial about the various options here.

Manishearth replied:
also it's not normal for build scripts to hit the network, and in particular people using non-cargo build systems _hate_ this (but it's been very abnormal in Rust even before that. People don't like nondeterminism in their build). I'm not sure we should be recommending this model, especially since most users will want to generate-once-and-check-in.

sffc replied:
CC @robertbastian 

Valid points. It _is_ super convenient though that databake-datagen is capable of being run in build.rs. There are examples of build.rs being _heavy_, so I'm not super worried about that; actually, this build.rs is fairly fast given a limited set of keys and locales. Being _non-deterministic_ is a bit trickier; at least worth a note that you probably don't want to do this unless you are a leaf crate.

Also, this model is already suggested in passing in data_management.md, and I think we've discussed it before. I'm not intending to introduce anything new here; just adding an example of something you can already do.

Manishearth replied:
Yeah I think we should _document_ this approach somewhere but I do not think we should be endorsing it too much. It's fine for now, perhaps we can clean this up in a followup.

robertbastian replied:
Well, the network only gets hit if you call `SourceData::with_{icuexport}_{tag,latest}`, which clearly document that they're downloading data. If you don't use those, datagen is offline.

Manishearth commented on docs/tutorials/cargo.md:
```diff
@@ -64,7 +64,9 @@ To generate your blob data, first run `cargo build` before you run `icu4x-datage
 ## Cargo.toml with Baked Provider
 
-If you wish to bake in data auto-generated in your build.rs, you need to add additional dependencies as follows:
+It is possible to auto-generate baked data in your build.rs. Use caution with this approach since it will make your build.rs file access the network and therefore be potentially non-deterministic; do not do this if you plan to have other people depending on your crate!
```
suggestion: we should not say that something is bad without providing an alternative: perhaps include a note that "if you would prefer to not deal with this problem; check in the generated files and regenerate as and when necessary"

Manishearth commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,119 @@
+
+This example has an additional section for auto-generating the data in build.rs. In your build.rs, invoke the ICU4X Datagen API with the set of keys you require. Don't worry; if using databake, you will get a compiler error if you don't specify enough keys.
+
+Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section and invoke `icu4x-datagen` manually. See [data_management.md](./data_management.md) for more information.
```
```suggestion
Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` sectio,  invoke `icu4x-datagen` manually, and potentially check the output in to your version control. See [data_management.md](./data_management.md) for more information.
```

robertbastian commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,119 @@
+publish = false
+
+[dependencies]
+icu = { version = "1.0", features = ["experimental"] }
```
Make sure to update this after your other PR

sffc replied:
Also requires 1.1

sffc replied:
Done

robertbastian commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,119 @@
+icu_provider_blob = "1.0"
+```
+
+To generate your blob data, first run `cargo build` before you run `icu4x-datagen`.
+
+For more information on building ICU4X data, including whether to check in the postcard file to your repository, see [data_management.md](./data_management.md).
````
Building before datagen is only required with the `keys-from-bin` option. Let's defer to the datagen tutorial.

```suggestion
To learn about building ICU4X data, including whether to check in the postcard file to your repository, see [data_management.md](./data_management.md).
```

robertbastian commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,119 @@
+[dependencies]
+icu = "1.0"
+icu_provider = "1.0" # for databake
+litemap = "0.6" # for databake
```
litemap will be removed with 1.1

sffc replied:
(requires 1.1)

robertbastian replied:
All our tutorials are being continuously updated, doing it all at release time is a lot of work. Can we somehow release them instead of serving them from head?

robertbastian replied:
bump

sffc replied:
Done

robertbastian commented on docs/tutorials/cargo_tests/baked/build.rs:
```diff
@@ -0,0 +1,27 @@
+use icu::locid::langid;
+use icu_datagen::CldrLocaleSubset;
+use icu_datagen::SourceData;
+use icu_provider::marker::KeyedDataMarker;
```
Maybe we should have a datagen prelude that reexports `langid` and `KeyedDataMarker` so that datagen can be called without two additional dependencies.

sffc replied:
https://github.com/unicode-org/icu4x/issues/2957

robertbastian replied:
You can now use it.

sffc replied:
Done

robertbastian commented on docs/tutorials/cargo_tests/baked/build.rs:
```diff
@@ -0,0 +1,27 @@
+        &[icu::plurals::provider::CardinalV1Marker::KEY],
+        &SourceData::default()
+            // Note: We use "latest" to stay fresh with the latest CLDR data.
+            .with_cldr_for_tag("latest", CldrLocaleSubset::Modern)
```
"latest" is not a valid version, `with_cldr_latest` resolves the "latest" GitHub tag to a version.

sffc replied:
Migrated to LATEST_TESTED_CLDR_TAG

robertbastian commented on docs/tutorials/cargo_tests/baked/build.rs:
```diff
@@ -0,0 +1,27 @@
+        vec![icu_datagen::Out::Module {
+            mod_directory: PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("baked_data"),
+            pretty: true,
+            insert_feature_gates: false,
```
Maybe add comments for these?

sffc replied:
It's hard to write comments that are as useful as the docs comments on Out::Module; they lack context, and good comments here would be just duplicating the docs. But I can add a link to the docs page.

robertbastian commented on docs/tutorials/cargo_tests/baked/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+use icu::plurals::PluralRules;
+
+mod baked {
+    include!("../baked_data/mod.rs");
```
Use the new style of declaring a struct and using the macro to add the impls.

sffc replied:
Can't do that until the new datagen is released. These examples are on crates.io code versions.

robertbastian replied:
.

sffc replied:
This is done.

robertbastian replied:
* I prefer defining the struct one level above the `include`
* I realize we're doing this in testdata to align with the constructors, but I'd rather call it `BakedProvider` than `UnstableProvider`. There's nothing unstable about the provider itself.

sffc replied:
done

robertbastian commented on docs/tutorials/cargo_tests/buffer/build_data.sh:
```diff
@@ -0,0 +1,18 @@
+    --format blob \
+    --keys-for-bin target/debug/buffer \
+    --locales my en-ZA \
+    --cldr-tag 42.0.0 \
```
`latest` for consistency?

sffc replied:
Done

robertbastian commented on docs/tutorials/data_management.md:
```diff
@@ -35,6 +35,23 @@ Let's dissect this invocation:
+
+You should check in the generated data if:
+
+1. You want fully reproducible builds
+2. You don't want any non-determinism
```
If you don't use downloads, builds are reproducible and deterministic.

sffc commented:
There's a lot of stuff here that is going to change after 1.1 is released. I'd therefore like to keep this PR open as the last change for 1.1, wait for it all to land in crates.io, then update this PR for 1.1 and merge it soon afterwards.

robertbastian commented:
1.1 is released, let's close the milestone.

sffc commented:
I started updating this PR but I discovered some issues in the 1.1 release:

- #3046
- #3047
- #3048

sffc commented:
The CI is passing with the workarounds in place:

https://github.com/sffc/omnicu/actions/runs/4050622606/jobs/6968179680

robertbastian commented on .github/workflows/release-test.yml:
```diff
@@ -0,0 +1,55 @@
+# This CI action runs the release crates as downloaded from crates.io. It does not block PRs
+# and it is okay if it is flaky from time to time.
+
+name: Release Build
```
I don't think this makes much sense. This runs on every push, but the release crates only change once every few months. Also, we'll only realise that we've broken an example *after the release*.

What I'd recommend doing instead:

* create a new workspace in `tutorials/cargo_tests`
  * (why `cargo_tests` and not `cargo` anyway?)
* use the [`patch` key](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section) on the workspace level to point all crates to local crates without specifying that in the examples themselves.
* If `tutorials/cargo_tests` itself changes, run a second pass with the patch key removed to test against crates.io

sffc replied:
Good idea! The catch with using `.cargo/config.toml` is that it's hard to unset the configurations; you need to do it one by one. So instead I created the config file, named it `patch.toml`, and when you want to use it, you pass it with the `--config` CLI option. I made a cargo-make job that applies the patch and hooked it up to run in CI.

robertbastian commented on Cargo.toml:
```diff
@@ -61,11 +61,7 @@ members = [
-    "docs/tutorials/cargo_tests/experimental",
-    "docs/tutorials/cargo_tests/lazy_static",
-    "docs/tutorials/cargo_tests/testdata",
+    "docs/tutorials/cargo_tests",
```
I think other workspaces are automatically excluded. We don't exclude the fuzzing crates or `icu_capi_staticlib_tiny` that declare they're a workspace.

sffc replied:
Hmm, maybe. I don't want to get messy with global feature resolution so I'll add a comment saying this is maybe not necessary

robertbastian replied:
Let's try to be consistent and remove this or add all non-workspace crates

sffc replied:
ok

robertbastian commented on docs/tutorials/cargo_tests/baked/Cargo.toml:
```diff
@@ -7,7 +7,7 @@
 [package]
-name = "demo"
+name = "demo_baked"
 version = "0.1.0"
```
Nit: we bump version numbers before publishing, so new crates should start at `0.0.0`. Also all our no-publish crates are set to `0.0.0`.

robertbastian commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,117 @@
+The most basic Cargo.toml to get you off the ground is the following:
+
+```toml
+[package]
+name = "demo_testdata"
+version = "0.1.0"
+edition = "2021"
+publish = false
````
I don't think these keys are very relevant to the example, given that you also link to the actual `Cargo.toml`.

sffc replied:
Do you suggest I remove the entire `[package]` section, or only the `publish =`?

robertbastian commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,117 @@
+icu_testdata = "1.1"
+```
+
+In your main.rs, you can use most functionality of ICU4X.
````
'most' is pretty handwavy. You can use all stable functionality, for some locales (link to test locales?).

sffc replied:
You also can't use certain features that are not re-exported, like most of the data provider crates.

robertbastian commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,117 @@
+edition = "2021"
+publish = false
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
remove

robertbastian commented on docs/tutorials/cargo_tests/baked/README.md:
```diff
@@ -0,0 +1,9 @@
+# demo [![crates.io](https://img.shields.io/crates/v/demo)](https://crates.io/crates/demo)
```
Does it make sense to generate READMEs for no-publish crates?

sffc replied:
Well, I made the README have a useful link to cargo.md, and it's more work to exclude these than it is to just make them partially useful.

robertbastian commented on docs/tutorials/cargo_tests/lazy_static/src/main.rs:
```diff
@@ -0,0 +1,31 @@
+}
+
+fn main() {
+    // TODO(#3048): This panics, but it shouldn't
```
fixed.

sffc replied:
Done

robertbastian commented on tools/make/tests.toml:
```diff
@@ -31,3 +31,23 @@ description = "Run all Rust doctests with all features"
+pwd = pwd
+set_env CARGO_TARGET_DIR ${pwd}/target
+
+exec --fail-on-error cargo run --release --manifest-path=baked/Cargo.toml
```
You can use `-p demo_baked` syntax as it's in-workspace.

sffc replied:
Done

sffc commented:
Thanks for the comments, but I haven't finished addressing feedback. I pushed a WIP last Friday. I'll hit the re-review button when ready.

sffc commented on docs/tutorials/cargo_tests/buffer/run.sh:
```diff
@@ -0,0 +1,8 @@
+# This file is part of ICU4X. For terms of use, please see the file
```
Yeah this is obsolete. Good catch. Removed.

sffc commented:
I think I've resolved everything except for wanting to try using the Cargo OUT_DIR variable for the baked data.

robertbastian commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,84 @@
+
+This example has an additional section for auto-generating the data in build.rs. In your build.rs, invoke the ICU4X Datagen API with the set of keys you require. Don't worry; if using databake, you will get a compiler error if you don't specify enough keys.
+
+Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section,  invoke `icu4x-datagen` manually, and potentially check in the output to your version control. See [data_management.md](./data_management.md) for more information.
```
```suggestion
Use caution with the build.rs approach since it will make your build.rs file access the network and therefore be potentially non-deterministic. As an alternative, you can remove the `build-dependencies` section,  invoke `icu4x-datagen` manually, and check in the output to your version control. See [data_management.md](./data_management.md) for more information.
```

robertbastian commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,84 @@
+lazy_static = "1.4"
+```
+
+You can now, for example, persist a particular ICU4X object in a [lazy_static](https://docs.rs/lazy_static/latest/lazy_static/).
````
```suggestion
You can now use most ICU4X types when `Send + Sync` are required, for example to persist them in a [lazy_static](https://docs.rs/lazy_static/latest/lazy_static/).
```

robertbastian commented on docs/tutorials/cargo_tests/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+    "lazy_static",
+    "testdata",
+]
+exclude = [
```
nit: remove

sffc replied:
ok

robertbastian commented on docs/tutorials/cargo_tests/buffer/Makefile:
```diff
@@ -0,0 +1,25 @@
+	echo "Target dir: $(CARGO_TARGET_DIR)"
+	cargo install icu_datagen --root .
+
+bin/demo_buffer: Cargo.toml src/*.rs
```
This doesn't capture dependency changes. Use `FORCE` to let cargo figure out if it needs to rebuild.

You should then remove the `cp` below and have the postcard rule use the binary from the target dir directly. Otherwise the mtime will change too much.

sffc replied:
I can use `FORCE` and then `cp -p` to not touch the mtime when copying. I still need the `cp` because I can't point the next task to depend on a file in an arbitrary directory (CARGO_TARGET_DIR).

sffc replied:
Also I'm removing the `cargo build --release` from the Makefile because the actual invocation of that command could vary (such as the cargo config file)

robertbastian replied:
Hmm, that's a downside of passing the config file in the command. You could put it back in `.cargo/config.toml` and in the crates.io CI rename it to `config.toml.disabled` to disable it.

sffc replied:
Strongly prefer a solution that doesn't involve changing the filesystem.

robertbastian replied:
can you disable it on the command line? `--config patch = []` or `--config.patch=[]`?

robertbastian commented on tools/make/tests.toml:
```diff
@@ -31,3 +31,47 @@ description = "Run all Rust doctests with all features"
+pwd = pwd
+set_env CARGO_TARGET_DIR ${pwd}/target
+
+# Note: This uses the --release profile to distinguish build artifacts from the local job
```
mmh do we plan on running this in CI?

sffc replied:
No. I added it to release.md. Also, I deleted this comment because it was wrong.

robertbastian replied:
But now yes? The only thing that should be built in release mode is the binary we pass to datagen. Everything else can run in debug mode.

sffc replied:
In most cases we're building only 1 crate. The only one where we run 2 builds is when installing icu4x-datagen, which has a separate discussion thread. I think this thread is therefore not actionable.

robertbastian commented on tools/make/tidy.toml:
```diff
@@ -107,7 +107,7 @@ script = '''
 
 glob_pattern = set "./**/Cargo.toml"
-glob_skip_pattern = array "ffi/gn/vendor/**/*" "ffi/gn/third_party_tools/**/*" "ffi/diplomat/*/examples/**/*" "tools/**/*" "**/fuzz/*" "**/target/**/*"
+glob_skip_pattern = array "ffi/gn/vendor/**/*" "ffi/gn/third_party_tools/**/*" "ffi/diplomat/*/examples/**/*" "tools/**/*" "**/fuzz/*" "**/target/**/*" "docs/tutorials/cargo_tests/Cargo.toml"
```
thought: we should teach these scripts about `publish = false` and `[workspace]`

robertbastian commented on tools/make/tidy.toml:
```diff
@@ -175,7 +175,7 @@ exit_on_error true
 glob_pattern = set "./**/Cargo.toml"
 license = canonicalize LICENSE
-skip_path_prefixes = array "ffi/gn/vendor/" "ffi/gn/third_party_tools/"
+skip_path_prefixes = array "ffi/gn/vendor/" "ffi/gn/third_party_tools/" "ffi/diplomat/c/examples" "docs/tutorials/cargo_tests/"
```
remove `ffi/diplomat/c/examples/fixeddecimal_tiny/icu_capi_staticlib_tiny/LICENSE`

sffc replied:
Actually everything in `ffi/diplomat/c/examples` is already being excluded so I went ahead and deleted LICENSE

robertbastian commented on .github/workflows/release-test.yml:
```diff
@@ -0,0 +1,55 @@
+
+jobs:
+  test-cargo-tutorial:
+    name: "Test Release via the Cargo tutorials"
```
nit: odd name

sffc replied:
improved

robertbastian commented on .github/workflows/release-test.yml:
```diff
@@ -0,0 +1,55 @@
+      uses: actions-rs/cargo@v1.0.1
+      with:
+        command: make
+        args: test-cargo-tutorial
```
does this exist?

sffc replied:
fixed

robertbastian commented on docs/tutorials/cargo_tests/baked/build.rs:
```diff
@@ -0,0 +1,34 @@
+            // For more information on these options, see:
+            // <https://icu4x.unicode.org/doc/icu_datagen/enum.Out.html#variant.Module>
+            mod_directory,
+            pretty: true,
```
`pretty: true` requires `rustfmt`, which might trip up clients if they run this on CI. Let's use `false` in all examples.

sffc replied:
ok

robertbastian commented on docs/tutorials/cargo_tests/experimental/src/main.rs:
```diff
@@ -0,0 +1,19 @@
+    let segmenter = GraphemeClusterSegmenter::try_new_unstable(
+        &icu_testdata::unstable(),
+    )
+    .expect("Grapheme break data is present in testdata");
```
Nit: should be present, also in other files

sffc replied:
done

robertbastian replied:
how did this get reverted?

sffc replied:
not sure but I'm fixing it again.

sffc commented:
The "release" CI job runs in the repo context, here: https://github.com/sffc/omnicu/actions/runs/4168109135/jobs/7214515460

robertbastian commented on .github/workflows/build-test.yml:
```diff
@@ -116,6 +116,48 @@ jobs:
+  test-cargo:
+    strategy:
+      matrix:
+        os: [ ubuntu-latest, macos-latest, windows-latest ]
```
I don't think multi-OS adds much value. I don't want to add more macOS CI because the number of macOS workers is more limited than ubuntu.

sffc replied:
I was thinking about that but I think that this is one where we probably really do want to make sure multi-OS works, because this is testing how real developers will interact with the project, and real developers are more likely to be on these other OSes.

robertbastian replied:
Multi-OS works because
* We already run `cargo test` on all OSes
* We already test datagen on all OSes
* We can assume `cargo` works on all OSes

I don't think we have to test this, and in fact the buffer-example apparently is not even OS-agnostic.

We only have 5 concurrent macOS workers. Currently each PR uses two (test and testdata), if we add a third even two PRs will slow each other down.

sffc replied:
I think we don't thoroughly test datagen `--from-bin` on all OS on CI

Also please note that the latest version of this PR only runs the full test on a daily cronjob

Is there anything actionable here?

robertbastian commented on .github/workflows/release-test.yml:
```diff
@@ -13,8 +13,8 @@ on:
   push
```
I think this is a waste of CI resources, just look how many commits you have in this PR alone. Let's have this as a manually triggered job in the release checklist.

sffc replied:
I'm happy to remove this CI, but I don't care as much about CI resources that run on main or on repository branches.

I already added it as a manually triggered job in the release checklist.

sffc replied:
Anything actionable?

robertbastian commented on docs/tutorials/cargo_tests/buffer/Makefile:
```diff
@@ -6,14 +6,16 @@ export CARGO_TARGET_DIR ?= target
 	echo "Target dir: $(CARGO_TARGET_DIR)"
-	cargo build --release
-	cp "$(CARGO_TARGET_DIR)"/release/demo_buffer bin/demo_buffer
+	echo "If the following command fails, run cargo build --release first."
```
That's not very makefile-y.

`ls "$(CARGO_TARGET_DIR)"/release/demo_buffer || cargo build --release`

sffc replied:
eh

sffc replied:
Added more comments

robertbastian commented on tools/make/tests.toml:
```diff
@@ -44,13 +44,15 @@ rm -f Cargo.lock
+set_env CARGO "cargo --config=../patch.toml"
 
-exec --fail-on-error cargo --config patch.toml run -p demo_baked
+exec --fail-on-error cargo --config patch.toml run --release -p demo_baked
```
If you just want to check that these build, don't use release mode. It'll unnecessarily slow down CI.

sffc replied:
I need `--release` in order for data slicing to work

sffc replied:
I need at least one of them to be `--release` so it makes sense for all of them to be `--release`. It is not slow to build these in `--release`. Datagen is slow in `--release` because it has a large amount of dependencies so LTO takes forever.

robertbastian replied:
`--release` doesn't have debug symbols, so if any of these fail they'll be annoying to debug.

sffc replied:
okay I made the `-local` task run in debug and `-cratesio` run in release.

sffc commented on docs/tutorials/cargo_tests/baked/src/main.rs:
```diff
@@ -0,0 +1,27 @@
+
+//! This is a demo project using baked data.
+//! 
+//! For more information, see the tutorial [cargo.md](../../cargo.md).
```
It works from here:

https://github.com/unicode-org/icu4x/tree/615c2c623ca6682f43f6cb9c7be7dcf8f3750639/docs/tutorials/cargo_tests/baked

sffc commented:
OK, I added a single new job called "check-cargo" which runs in less than 5 minutes. It runs the build script, but not the postcard file. Meanwhile, I added the full job (which builds and runs everything) on a daily cron job.

sffc commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,100 @@
+
+name: Build Daily on Main
+
+on:
+  push:
+    branches: [ main ]
+  schedule:
+    # Run this CI every day at 13 UTC / 5AM PST / 2PM CET
+    - cron:  '0 13 * * *'
```
@Manishearth Thoughts on adding a new file to collect tasks that we only want to run once a day?

sffc replied:
This is at 13 and the other is at 14

Manishearth replied:
+1 go for it yolo

robertbastian replied:
You're running this on push and at 13 UTC?

sffc replied:
I moved it to 12 UTC and removed the running on main merge

robertbastian commented on docs/tutorials/cargo_tests/baked/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+##########################################
+
+[package]
+name = "demo_baked"
```
Naming. Currently there's
* "cargo tutorial" in text, make jobs
* "cargo" in CI jobs, make jobs
* "cargo_tests" in the directory name
* "demo_foo" in the crate names

Can we align everything to tutorial?
* `ci-job-foo-cargo-tutorial-bar`
* `docs/tutorials/cargo`
* crate name `baked_tutorial`/`tutorial_baked`
  * Or just `baked`, as it's in the tutorial workspace?

sffc replied:
did `tutorial_*`

robertbastian replied:
What about the `cargo_tests` directory name?

sffc replied:
Not sure about the directory name. `cargo_tests` is probably not right, but I don't like just `cargo` because then the directory looks too similar to `.cargo` which has real meaning. I could do

- `crates`
- `cargotoml`
- `cargo_examples`
- `examples`

robertbastian replied:
I like `crates` or `cargo_tutorials`, although that's a bit repetitive. I'm happy with `examples` if you replace all the `-tutorials` with `-examples` (which I don't expect you to) 🙃

robertbastian commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,100 @@
+        command: make
+        args: test-cargo-tutorial-cratesio
+
+  test-cargo-local:
```
```suggestion
  test-cargo-tutorial-local:
```

sffc replied:
ok

robertbastian commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,100 @@
+    - cron:  '0 13 * * *'
+
+jobs:
+  test-cargo-cratesio:
```
```suggestion
  test-cargo-tutorial-cratesio:
```

sffc replied:
ok

robertbastian commented on .github/workflows/build-test.yml:
```diff
@@ -116,6 +116,44 @@ jobs:
         args: ci-job-test
 
+  # ci-job-test-cargo
+  check-cargo:
```
```suggestion
  check-cargo-tutorial:
```

sffc replied:
ok

robertbastian commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,100 @@
+    - name: Show the selected Rust toolchain
+      run: rustup show
+
+    # Actual job
```
Jobs that run datagen should load the source data cache.

sffc replied:
ok

sffc commented:
Note: We can add `SourceData::Baked` which is exhaustive but has an `options` non-exhaustive field.

sffc commented on Makefile.toml:
```diff
@@ -65,6 +65,13 @@ dependencies = [
     "test-all-features",
 ]
 
+[tasks.ci-job-check-cargo]
```
done

sffc commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,100 @@
+        command: make
+        args: test-cargo-tutorial-cratesio
+
+  test-cargo-local:
```
It takes like 15 minutes. There's not much reason to run it because if it passes `cargo check`, it _probably_ runs all the way through. It's a lot of CI cycles for not much benefit. But I can add it back to build-test.yml if you prefer

robertbastian replied:
I'd prefer that.

sffc replied:
I changed the job back to the full build and changed it to debug mode. I'm also seeing whether it might be faster if we attach it to the end of one of the existing jobs.

sffc replied:
Runtimes are now:

- 8m 10s for a standalone job (1m 21s for setup and 6m 49s for the job)
- 5m 0s tacked onto the end of the testdata job
- 6m 14s tacked onto the end of the test job

I am therefore inclined to keep this as its own job.

robertbastian replied:
I wonder where those gains come from, it's a separate workspace so everything should be rebuilt.

sffc replied:
I believe the gains are from re-using the top level target directory to build a debug datagen binary, instead of cargo installing it.

sffc commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,84 @@
+
+```toml
+[dependencies]
+icu = { version = "1.1", features = ["experimental"] }
````
Yes it's fixed in the real Cargo.toml file but I forgot to fix it here. Done

sffc commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,84 @@
+icu_testdata = { version = "1.1", features = ["icu_segmenter"] }
+```
+
+In your main.rs, you can now use experimental features.
````
done

sffc commented on docs/tutorials/cargo.md:
````diff
@@ -0,0 +1,84 @@
+icu_provider_blob = "1.1"
+```
+
+To learn about building ICU4X data, including whether to check in the postcard file to your repository, see [data_management.md](./data_management.md).
````
done

sffc commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,84 @@
+[build-dependencies]
+icu = "1.1"
+icu_datagen = "1.1"
+icu_provider = "1.1"
```
done

sffc commented on docs/tutorials/cargo.md:
```diff
@@ -0,0 +1,84 @@
+icu = "1.1"
+icu_testdata = "1.1"
+icu_provider = { version = "1.1", features = ["sync"] }
+lazy_static = "1.4"
```
ok I can remove it from here and leave it in the real Cargo.toml

sffc commented on docs/tutorials/cargo_tests/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# These demos should always build with the freshest dependencies
```
ok.

sffc commented on docs/tutorials/cargo_tests/baked/build.rs:
```diff
@@ -0,0 +1,37 @@
+
+    icu_datagen::datagen(
+        Some(&[langid!("ru")]),
+        &[icu::plurals::provider::CardinalV1Marker::KEY],
```
I want to keep the small set for build times. Added a comment

robertbastian commented on docs/tutorials/cargo_tests/baked/build.rs:
```diff
@@ -0,0 +1,35 @@
+        &[icu::plurals::provider::CardinalV1Marker::KEY],
+        &SourceData::default()
+            .with_cldr_for_tag(SourceData::LATEST_TESTED_CLDR_TAG, CldrLocaleSubset::Modern)
+            .expect("Source data should download successfully"),
```
this doesn't do any downloading, so that message is not accurate

sffc replied:
I guess the function is infallible now? This is probably a leftover from when I wrote this file against 1.0. I added this to https://github.com/unicode-org/icu4x/issues/2856

robertbastian replied:
It is, thanks for adding

sffc commented on tools/make/tests.toml:
```diff
@@ -31,3 +31,75 @@ description = "Run all Rust doctests with all features"
+
+# Build the postcard file
+exec --fail-on-error cargo --config patch.toml build --release -p demo_buffer
+exec --fail-on-error cargo --config patch.toml install --path ../../../provider/datagen --root .
```
done

robertbastian commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,112 @@
+
+    # Actual job
+    - name: Run `cargo make test-cargo-tutorial`
+      uses: actions-rs/cargo@v1.0.1
```
https://github.com/unicode-org/icu4x/pull/3101

sffc replied:
done here and elsewhere

robertbastian commented on .github/workflows/build-on-main.yml:
```diff
@@ -0,0 +1,112 @@
+        command: make
+        args: test-cargo-tutorial-cratesio
+
+  test-cargo-tutorial-local:
```
As discussed this is already in the other file.

sffc replied:
done

robertbastian commented on tools/make/tests.toml:
```diff
@@ -31,3 +31,80 @@ description = "Run all Rust doctests with all features"
+set_env CARGO_TARGET_DIR ${pwd}/target
+
+cd docs/tutorials/cargo_tests
+rm -f Cargo.lock
```
I think for local we didn't want this? There's no point checking in the file and then always deleting it

sffc replied:
done

robertbastian commented on tools/make/tests.toml:
```diff
@@ -31,3 +31,80 @@ description = "Run all Rust doctests with all features"
+exec --fail-on-error cargo build -p icu_datagen
+
+cd docs/tutorials/cargo_tests
+rm -f Cargo.lock
```
.

sffc replied:
done

robertbastian commented on .github/workflows/build-test.yml:
```diff
@@ -110,6 +110,44 @@ jobs:
+      if: steps.cargo-make-cache.outputs.cache-hit != 'true'
+      run: cargo +stable install cargo-make
+
+
```
I think this uses datagen now, so load the source data cache.

sffc replied:
yep

robertbastian commented on docs/tutorials/cargo_tests/.gitignore:
```diff
@@ -0,0 +1,3 @@
+bin/
+.crates.toml
```
what's this?

sffc replied:
They were files that cargo was making when it downloaded stuff from crates.io; they were empty and then they appeared to contain hashes of the crates or something. I didn't look too closely.

robertbastian commented on docs/tutorials/cargo_tests/buffer/Makefile:
```diff
@@ -0,0 +1,31 @@
+
+.PHONY: clean
+clean:
+	rm -f bin/icu4x-datagen
```
nit: `git clean -xf *` to stay consistent with gitignores

sffc replied:
ok

robertbastian commented on docs/tutorials/cargo_tests/experimental/src/main.rs:
```diff
@@ -0,0 +1,19 @@
+    let segmenter = GraphemeClusterSegmenter::try_new_unstable(
+        &icu_testdata::unstable(),
+    )
+    .expect("Grapheme break data should be present in testdata");
```
thought: it's unfortunate that the key being present in the data doesn't mean anything for `und`-only data

robertbastian reviewed (changes requested):
lgtm overall, some nits and the CI change we discussed
